src/librustc_mir/interpret/intern.rs

```diff
@@ -396,13 +420,13 @@ where
                 }
             }
         } else if ecx.memory.dead_alloc_map.contains_key(&alloc_id) {
-            // dangling pointer
-            throw_ub_format!("encountered dangling pointer in final constant")
+            // Codegen does not like dangling pointers, and generally `tcx` assumes that
+            // all allocations referenced anywhere actually exist. So, make sure we error here.
+            ecx.tcx.sess.span_err(ecx.tcx.span, "encountered dangling pointer in final constant");
```

Comment:
I assume it is too late here to adjust the allocations containing the dangling pointer? Because one thing we could do is replace all pointers to it with a (dangling) integer pointer (taking the alignment of the allocation as its base address). That would maintain data structure consistency (no dangling `AllocId` in `tcx`) and avoid having to hard-error here.

Reply:
We could pull the `remove` invocation upfront to where elements are added to `todo` and put `(AllocId, Allocation)` pairs into `todo`. Then we'll only have to figure out how to make the visitor allow mutating the value in place. Likely that will require a sort of `MutVisitor`/`Visitor` duplication like `mir::Visitor` has.

Simpler would probably be to do a first pass over an Allocation and replace all dangling `AllocId`s in its relocation with integer pointers, but then we have a hard time telling apart user's integer pointers from our synthetic ones.

So the only thing that I can think of is to make this site a `delay_span_bug` and to do the "replace all dangling AllocIds` pass just before invoking `intern_const_alloc`

Reply:
I don't think we can make this a `delay_span_bug` here, dangling raw pointers would not be a "bug" if we have synthetic ones.

I guess the weirdest part about the synthetic pointers is that inequal pointers could become equal, which doesn't seem great. Maybe we should just synthesize 1-byte global allocations and make them point to that?

Anyway I think that is something which can be resolved later, after landing this PR.

Reply:
yes

Reply:
> I guess the weirdest part about the synthetic pointers is that inequal pointers could become equal, which doesn't seem great. Maybe we should just synthesize 1-byte global allocations and make them point to that?

I'd prefer to just use integer pointers. There's no way their identity is important (if they were pointing to a static, they wouldn't be dangling), so deduplication is totally permissible.

Reply:
Opened an issue for this: https://github.com/rust-lang/rust/issues/72215